### PR TITLE
Move generated files out of project build directories

### PR DESCRIPTION
**Pull request by kageiit**
This moves out merged manifest, proguard and keystore generated files out of project build dirs into the root `.okbuck/gen` folder.
The main motivation for this is to improve the performance of watchman queries as the `ignore_dirs` [option](https://facebook.github.io/watchman/docs/config.html#ignoredirs) only applies to full paths and not  patterns.
This was causing a lot of churn in watchman and delay in starting up buck after okbuck in large projects when the generated files are put in the project build dirs.
A lot of logic around generated files is also simplified and stale gen files are now cleaned up properly by the `okbuckClean` task

### CODE REVIEW

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/base/BuckRuleComposer.java [hunk 1]
```diff
@@ -25,11 +25,14 @@
...
+        if (file == null) {
+            return null;
+        }
+        String ext = file;
+        int ind = FilenameUtils.indexOfLastSeparator(file);
         if (ind >= 0) {
-            ext = new StringBuilder(dep).replace(ind, ind + 1, ":").toString();
+            ext = new StringBuilder(file).replace(ind, ind + 1, ":").toString();
         }
         return "//" + ext;
```
Comment (raviagarwal7) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/base/BuckRuleComposer.java]: nit: `String ext = new StringBuilder(file)` and use it to prepend `//`

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/base/BuckRuleComposer.java [hunk 2]
```diff
@@ -25,11 +25,14 @@
                 .collect(Collectors.toSet());
     }
 
-    public static String external(final String dep) {
-        String ext = dep;
-        int ind = FilenameUtils.indexOfLastSeparator(dep);
+    public static String fileRule(final String file) {
```
Comment (raviagarwal7) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/base/BuckRuleComposer.java]: nit: filepath

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 3]
```diff
@@ -397,6 +406,10 @@ abstract class AndroidTarget extends JavaLibTarget {
         return expandedConfigs
     }
 
+    File getGenPath(String... paths) {
```
Comment (raviagarwal7) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: `getGenFile`?

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 4]
```diff
@@ -80,6 +84,11 @@ abstract class AndroidTarget extends JavaLibTarget {
             minSdk = baseVariant.mergedFlavor.minSdkVersion.apiLevel
             targetSdk = baseVariant.mergedFlavor.targetSdkVersion.apiLevel
         }
+
+        // Create gen dir
+        genDir = [OkBuckGradlePlugin.OKBUCK_GEN, path, name].join(File.separator)
```
Comment (raviagarwal7) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: probably use path api to resolve the paths.

Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: Sure

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy (hunk 3 above)
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: We use this for folder as well so I called it path

**Comment by kageiit** [comment]
The lint regression test caught a bug. Thanks @clonetwin26 !

**Comment by kageiit** [comment]
Unit tests are failing. Need to merge https://github.com/facebook/buck/pull/1282 before proceeding with this change

**Pull request merged**